lib/StackAnalysis/SegregateDirectStackAccesses.cpp

```diff
@@ -0,0 +1,152 @@
+/// \file SegregateDirectStackAccesses.cpp
+/// \brief Segregate direct stack accesses from all other memory accesses
+/// through alias information. This provides a way to say that stack accesses do
+/// not interfere with any other memory access.
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/MDBuilder.h"
+#include "llvm/IR/PatternMatch.h"
+
+#include "revng/StackAnalysis/SegregateDirectStackAccesses.h"
+#include "revng/Support/Assert.h"
+
+using namespace llvm;
+using SDSAP = SegregateDirectStackAccessesPass;
+
+PreservedAnalyses SDSAP::run(Function &F, FunctionAnalysisManager &FAM) {
+  Context = &(F.getContext());
+
+  // Get the result of the GCBI analysis
+  auto &M = *F.getParent();
+  const auto &MAMProxy = FAM.getResult<ModuleAnalysisManagerFunctionProxy>(F);
+  GCBI = MAMProxy.getCachedResult<GeneratedCodeBasicInfoAnalysis>(M);
+  if (!GCBI)
+    GCBI = &(FAM.getResult<GeneratedCodeBasicInfoAnalysis>(F));
+  revng_assert(GCBI != nullptr);
+
+  // Populate the two buckets with all load and store instruction of the
+  // function, properly segregated.
+  segregateAccesses(F);
+
+  // Adorn IR with the alias information collected before.
+  decorateStackAccesses();
+
+  return PreservedAnalyses::none();
+}
+
+void SDSAP::segregateAccesses(Function &F) {
+  using namespace PatternMatch;
+  Value *LoadSP = nullptr;
+  bool Found = false;
+
+  for (BasicBlock &BB : F) {
+    for (Instruction &I : BB) {
+      if (auto *LI = dyn_cast<LoadInst>(&I)) {
+        if (GCBI->isSPReg(skipCasts(LI->getPointerOperand()))) {
+          LoadSP = LI;
+          Found = true;
```

Comment:
Mmmh, do we expect to have a single load from `sp`? If so, we should assert this.

Reply:
Uhmm, there are cases in which may not have a first load from `SP`, so we should not.

Reply:
I mean: what if we have multiple loads from `sp`? You seem to silently ignore such a situation.

Reply:
It really suffices to find the first load from `SP`.

Reply:
Below you're using `LoadPtr` to match IR. If there are multiple loads from `sp`, those matches will fail.
I don't think this is what we want.

Reply:
OK, checked better and re-run the tests, and if there exists a load from `SP`, it is really the unique one. Should we check this?

Reply:
Yeah, just `revng_assert(not Found)`.